File: CygniParser/src/cygni/util/AstJson.java
package cygni.util;

import cygni.ast.*;

import java.util.ArrayList;
import java.util.HashMap;

public class AstJson {
    public static String toString(Program program) {
        return JsonUtil.toString(makeList(program.nodes));
    }

    public static Object visit(Node node) {
        if (node instanceof BinaryOp) {
            return visitBinary((BinaryOp) node);
        } else if (node instanceof UnaryOp) {
            return visitUnary((UnaryOp) node);
        } else if (node instanceof Name) {
            return visitName((Name) node);
        } else if (node instanceof Constant) {
            return visitConstant((Constant) node);
        } else if (node instanceof IfThen) {
            return visitIfThen((IfThen) node);
        } else if (node instanceof IfElse) {
            return visitIfElse((IfElse) node);
        } else if (node instanceof Block) {
            return visitBlock((Block) node);
        } else if (node instanceof Assign) {
            return visitAssign((Assign) node);
        } else if (node instanceof Def) {
            return visitDef((Def) node);
        } else if (node instanceof Return) {
            return visitReturn((Return) node);
        } else if (node instanceof Call) {
            return visitCall((Call) node);
        } else if (node instanceof InitArray) {
            return visitInitArray((InitArray) node);
        } else if (node instanceof While) {
            return visitWhile((While) node);
        } else if (node instanceof Var) {
            return visitVar((Var) node);
        } else {
            return "(null: " + node + ")";
        }
    }

    public static HashMap<String, Object> makeMap(String key0, Object value0) {
        HashMap<String, Object> map = new HashMap<String, Object>();
        map.put(key0, value0);
        return map;
    }

    public static HashMap<String, Object> makeMap(String key0, Object value0, String key1, Object value1) {
        HashMap<String, Object> map = makeMap(key0, value0);
        map.put(key1, value1);
        return map;
    }

    public static HashMap<String, Object> makeMap(String key0, Object value0, String key1, Object value1, String key2, Object value2) {
        HashMap<String, Object> map = makeMap(key0, value0, key1, value1);
        map.put(key2, value2);
        return map;
    }

    public static HashMap<String, Object> makeMap(String key0, Object value0, String key1, Object value1,
                                                  String key2, Object value2, String key3, Object value3) {
        HashMap<String, Object> map = makeMap(key0, value0, key1, value1, key2, value2);
        map.put(key3, value3);
        return map;
    }

    public static HashMap<String, Object> makeMap(String key0, Object value0, String key1, Object value1,
                                                  String key2, Object value2, String key3, Object value3,
                                                  String key4, Object value4) {
        HashMap<String, Object> map = makeMap(key0, value0, key1, value1, key2, value2, key3, value3);
        map.put(key4, value4);
        return map;
    }

    public static ArrayList<Object> makeList(ArrayList<Node> nodes) {
        ArrayList<Object> objects = new ArrayList<Object>();
        for (Node node : nodes) {
            objects.add(visit(node));
        }
        return objects;
    }

    public static Object visitBinary(BinaryOp node) {
        Object left = visit(node.left);
        Object right = visit(node.right);
        return makeMap("kind", node.kind,
                "left", left, "right", right);
    }

    public static Object visitUnary(UnaryOp node) {
        Object operand = visit(node.operand);
        return makeMap("kind", node.kind, "operand", operand);
    }

    public static Object visitName(Name node) {
        return makeMap("kind", "Name", "name", node.name);
    }

    public static Object visitConstant(Constant node) {
        return makeMap("kind", "Constant", "value", node.value.toString());
    }

    public static Object visitIfThen(IfThen node) {
        return makeMap("kind", "IfThen", "condition", visit(node.condition),
                "ifTrue", visit(node.ifTrue));
    }

    public static Object visitIfElse(IfElse node) {
        return makeMap("kind", "IfElse", "condition", visit(node.condition),
                "ifTrue", visit(node.ifTrue), "ifFalse", visit(node.ifFalse));
    }

    public static Object visitBlock(Block node) {
        return makeMap("kind", "Block", "nodes", makeList(node.nodes));
    }

    public static Object visitAssign(Assign node) {
        return makeMap("kind", "Assign", "left", visit(node.left), "value", visit(node.value));
    }

    public static Object visitDef(Def node) {
        return makeMap("kind", "Def",
                "name", node.name,
                "type-parameters", node.unknownTypes,
                "parameters", node.parameters,
                "body", visit(node.body));
    }

    public static Object visitReturn(Return node) {
        return makeMap("kind", "Return", "value", visit(node.value));
    }

    public static Object visitCall(Call node) {
        return makeMap("kind", "Call",
                "function", visit(node.function), "argument", makeList(node.arguments));
    }

    public static Object visitInitArray(InitArray node) {
        return makeMap("kind", "InitArray",
                "elements", makeList(node.elements));
    }

    public static Object visitWhile(While node) {
        return makeMap("kind", "While", "condition", visit(node.condition),
                "body", visit(node.body));
    }

    public static Object visitVar(Var node) {
        return makeMap("kind", "Var", "name", node.name, "value", visit(node.value));
    }
}

File: Compiler/Compiler/AstJson.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace Compiler
{
    public class AstJson
    {
        public AstJson()
        {
        }
        public JsonObject Visit(Program program)
        {
            return new JsonMap(
            new Dictionary<string, JsonObject>()
            {
                {"path", new JsonValue(program.path)},
                {"modules", new JsonArray(program.modules.Select(VisitNode).ToList()) },
                {"classes", new JsonArray(program.classes.Select(VisitNode).ToList()) },
            });
        }
        JsonObject VisitNode(Ast node)
        {
            switch (node.kind)
            {
                case Kind.Add:
                case Kind.Subtract:
                case Kind.Multiply:
                case Kind.Divide:
                case Kind.Modulo:
                case Kind.GreaterThan:
                case Kind.LessThan:
                case Kind.GreaterThanOrEqual:
                case Kind.LessThanOrEqual:
                case Kind.Equal:
                case Kind.NotEqual:
                    return VisitBinary((Binary)node);
                case Kind.And:
                    throw new NotImplementedException();
                case Kind.Or:
                    throw new NotImplementedException();
                case Kind.Not:
                    throw new NotImplementedException();
                case Kind.UnaryPlus:
                case Kind.UnaryMinus:
                    return VisitUnary((Unary)node);
                case Kind.IfThen:
                    return VisitIfThen((IfThen)node);
                case Kind.IfElse:
                    return VisitIfElse((IfElse)node);
                case Kind.Constant:
                    return VisitConstant((Constant)node);
                case Kind.Block:
                    return VisitBlock((Block)node);
                case Kind.Name:
                    return VisitName((Name)node);
                case Kind.Return:
                    return VisitReturn((Return)node);
                case Kind.Var:
                    return VisitVar((Var)node);
                case Kind.Def:
                    return VisitDef((Def)node);
                case Kind.Assign:
                    return VisitBinary((Binary)node);
                case Kind.Call:
                    return VisitCall((Call)node);
                case Kind.While:
                    return VisitWhile((While)node);
                case Kind.DefClass:
                    return VisitClass((DefClass)node);
                case Kind.DefModule:
                    return VisitModule((DefModule)node);
                case Kind.TypeSpecifier:
                    return VisitType((TypeSpecifier)node);
                case Kind.MemberAccess:
                    return VisitMemberAccess((MemberAccess)node);
                default:
                    throw new NotSupportedException();
            }
        }

        JsonObject VisitBinary(Binary node)
        {
            return new JsonMap(
            new Dictionary<string, JsonObject>()
            {
                {"id", new JsonValue(node.id.ToString())},
                {"kind", new JsonValue(node.kind.ToString())},
                {"left", VisitNode(node.left)},
                {"right", VisitNode(node.right)},
            });
        }

        JsonObject VisitUnary(Unary node)
        {
            return new JsonMap(
            new Dictionary<string, JsonObject>()
            {
                {"id", new JsonValue(node.id.ToString())},
                {"kind", new JsonValue(node.kind.ToString())},
                {"operand", VisitNode(node.operand)}
            });
        }

        JsonObject VisitType(TypeSpecifier node)
        {
            return new JsonMap(
            new Dictionary<string, JsonObject>()
            {
                {"id", new JsonValue(node.id.ToString())},
                {"kind", new JsonValue(node.kind.ToString())},
                {"name", new JsonValue(node.name)},
                {"parameters", new JsonArray(node.parameters.Select(VisitType).ToList())}
            });
        }

        JsonObject VisitDef(Def node)
        {
            return new JsonMap(
            new Dictionary<string, JsonObject>()
            {
                {"id", new JsonValue(node.id.ToString())},
                {"kind", new JsonValue(node.kind.ToString())},
                {"name", new JsonValue(node.name)},
                {"body", VisitNode(node.body)},
                {"type", VisitType(node.type)}
            });
        }

        JsonObject VisitClass(DefClass node)
        {
            return new JsonMap(
            new Dictionary<string, JsonObject>()
            {
                {"id", new JsonValue(node.id.ToString())},
                {"kind", new JsonValue(node.kind.ToString())},
                {"name", new JsonValue(node.name)},
                {"fields", new JsonArray(node.fields.Select(VisitNode).ToList())},
                {"methods", new JsonArray(node.methods.Select(VisitNode).ToList())}
            });
        }

        JsonObject VisitModule(DefModule node)
        {
            return new JsonMap(
     new Dictionary<string, JsonObject>()
     {
                {"id", new JsonValue(node.id.ToString())},
                {"kind", new JsonValue(node.kind.ToString())},
                {"name", new JsonValue(node.name)},
                {"fields", new JsonArray(node.fields.Select(VisitNode).ToList())},
                {"methods", new JsonArray(node.methods.Select(VisitNode).ToList())}
     });
        }

        JsonObject VisitCall(Call node)
        {
            return new JsonMap(
                 new Dictionary<string, JsonObject>()
                {
                {"id", new JsonValue(node.id.ToString())},
                {"kind", new JsonValue(node.kind.ToString())},
                {"function", VisitNode(node.function)},
                {"arguments", new JsonArray(node.arguments.Select(VisitNode).ToList())}
                });
        }

        JsonObject VisitIfThen(IfThen node)
        {
            return new JsonMap(
            new Dictionary<string, JsonObject>()
            {
                {"id", new JsonValue(node.id.ToString())},
                {"kind", new JsonValue(node.kind.ToString())},
                { "condition", VisitNode(node.condition) },
                { "ifTrue",VisitNode(node.ifTrue) }
            });
        }

        JsonObject VisitIfElse(IfElse node)
        {
            return new JsonMap(
            new Dictionary<string, JsonObject>()
            {
                {"id", new JsonValue(node.id.ToString())},
                {"kind", new JsonValue(node.kind.ToString())},
                { "condition", VisitNode(node.condition) },
                { "ifTrue",VisitNode(node.ifTrue) },
                { "ifFalse",VisitNode(node.ifFalse) }
            });
        }

        JsonObject VisitConstant(Constant node)
        {
            return new JsonMap(
            new Dictionary<string, JsonObject>()
            {
                {"id", new JsonValue(node.id.ToString())},
                {"kind", new JsonValue(node.kind.ToString())},
                {"type", new JsonValue(node.value.GetType().ToString())},
                {"value", new JsonValue(node.value.ToString())}
            });
        }

        JsonObject VisitBlock(Block node)
        {
            return new JsonMap(
            new Dictionary<string, JsonObject>()
            {
                {"id", new JsonValue(node.id.ToString())},
                {"kind", new JsonValue(node.kind.ToString())},
                {"expressions",new JsonArray( node.expressions.Select(VisitNode).ToList()) }
            });
        }

        JsonObject VisitName(Name node)
        {
            return new JsonMap(
            new Dictionary<string, JsonObject>()
            {
                {"id", new JsonValue(node.id.ToString())},
                {"kind", new JsonValue(node.kind.ToString())},
                {"name", new JsonValue(node.name)}
            });
        }

        JsonObject VisitReturn(Return node)
        {
            return new JsonMap(
            new Dictionary<string, JsonObject>()
            {
                {"id", new JsonValue(node.id.ToString())},
                {"kind", new JsonValue(node.kind.ToString())},
                {"value", VisitNode(node.value) }
            });
        }

        JsonObject VisitVar(Var node)
        {
            return new JsonMap(
            new Dictionary<string, JsonObject>()
            {
                {"id", new JsonValue(node.id.ToString())},
                {"kind", new JsonValue(node.kind.ToString())},
                {"type", node.type == null?new JsonValue("<NULL>"):VisitType(node.type) },
                {"value", node.value == null?new JsonValue("<NULL>"):VisitNode(node.value) }
            });
        }

        JsonObject VisitWhile(While node)
        {
            return new JsonMap(
            new Dictionary<string, JsonObject>()
            {
                {"id", new JsonValue(node.id.ToString())},
                {"kind", new JsonValue(node.kind.ToString())},
                {"condition", VisitNode(node.condition)},
                {"body", VisitNode(node.body)}
            });
        }

        JsonObject VisitMemberAccess(MemberAccess node)
        {
            return new JsonMap(
                new Dictionary<string, JsonObject>
                {
                    { "id", new JsonValue(node.id.ToString())},
                    { "kind", new JsonValue(node.kind.ToString())},
                    { "expression", VisitNode(node.expression)},
                    { "member",new JsonValue(node.member.ToString())}
                }
            );
        }
    }
}
